api/admin/settings.go

```diff
@@ -0,0 +1,221 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package admin
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/sirupsen/logrus"
+
+	"github.com/go-vela/server/api/types/settings"
+	"github.com/go-vela/server/compiler/native"
+	"github.com/go-vela/server/database"
+	"github.com/go-vela/server/queue"
+	cliMiddleware "github.com/go-vela/server/router/middleware/cli"
+	sMiddleware "github.com/go-vela/server/router/middleware/settings"
+	uMiddleware "github.com/go-vela/server/router/middleware/user"
+	"github.com/go-vela/server/util"
+)
+
+// swagger:operation GET /api/v1/admin/settings admin GetSettings
+//
+// Get the currently configured settings.
+//
+// ---
+// produces:
+// - application/json
+// security:
+//   - ApiKeyAuth: []
+// responses:
+//   '200':
+//     description: Successfully retrieved settings from the database
+//     schema:
+//       type: array
+//       items:
+//         "$ref": "#/definitions/Platform"
+//   '500':
+//     description: Unable to retrieve settings from the database
+//     schema:
+//       "$ref": "#/definitions/Error"
+
+// GetSettings represents the API handler to
+// captures settings stored in the database.
+func GetSettings(c *gin.Context) {
+	// capture middleware values
+	s := sMiddleware.FromContext(c)
+
+	logrus.Info("Admin: reading settings")
+
+	c.JSON(http.StatusOK, s)
+}
+
+// swagger:operation PUT /api/v1/admin/settings admin UpdateSettings
+//
+// Update the settings singleton in the database.
+//
+// ---
+// produces:
+// - application/json
+// parameters:
+// - in: body
+//   name: body
+//   description: Payload containing settings to update
+//   required: true
+//   schema:
+//     "$ref": "#/definitions/Platform"
+// security:
+//   - ApiKeyAuth: []
+// responses:
+//   '200':
+//     description: Successfully updated the settings in the database
+//     schema:
+//       "$ref": "#/definitions/Platform"
+//   '404':
+//     description: Unable to update the settings in the database
+//     schema:
+//       "$ref": "#/definitions/Error"
+//   '501':
+//     description: Unable to update the settings in the database
+//     schema:
+//       "$ref": "#/definitions/Error"
+
+// UpdateSettings represents the API handler to
+// update the settings singleton stored in the database.
+func UpdateSettings(c *gin.Context) {
+	// capture middleware values
+	s := sMiddleware.FromContext(c)
+	u := uMiddleware.FromContext(c)
+	ctx := c.Request.Context()
+
+	logrus.Info("Admin: updating settings")
+
+	// capture body from API request
+	input := new(settings.Platform)
+
+	err := c.Bind(input)
+	if err != nil {
+		retErr := fmt.Errorf("unable to decode JSON for settings: %w", err)
+
+		util.HandleError(c, http.StatusBadRequest, retErr)
+
+		return
+	}
+
+	if input.Compiler != nil {
+		if input.CloneImage != nil {
+			s.SetCloneImage(*input.CloneImage)
+		}
```

Comment:
i'd hate to bring in the docker lib, but some way to validate that the image you are updating exists would be nice, since it could cause a decent amount of breakage.

Reply:
sure i can look into that. it should be added to the server bootup as well to check the environment-provided value also, not just here, right?

Reply:
taken care of, see https://github.com/go-vela/server/pull/1110/commits/197e42ff5919269bd98ec8fb21e148d028131cbc

Reply:
lol..... okay this actually turned into a bigger effort because of this test:
https://github.com/go-vela/server/blob/8a744e48d224197478bb48a468186783b04c44db/compiler/native/compile_test.go#L1852

that is failing because there is no clone image provided, NOT because the yaml has no steps or stages.
the steps validation injects clone+init and doesnt actually fail but the test _thinks_ its failing, its actually failing because theres no clone image, so when i provide a clone image to get a valid compiler the test starts failing. 
its a not-my-problem-at-all scope creep where i need to now fix steps validation....
i dont feel comfortable changing compiler yaml validation in this PR

Reply:
im thinking of how i can get around this but now i know this test is a false positive so im sorta stuck.

fix now? no, not touching compiler functionality in this PR.
skip the test? gross, we might not ever get around to actually fixing it.
remove docker image validation for the compiler? i want to keep it.

ill think about it over lunch

Reply:
fwiw this lets us move on, but it hides the underlying issue. an issue that isnt at-all related to this PR
https://github.com/go-vela/server/pull/1110/commits/a43546e83ddd47c1da398c5e40967eea3d4cde9f#diff-a25c05febe73513fec1340739f5f59b78fb50a646dccf716b5825cd5fe3d8652R1849-R1851

Reply:
ah, interesting. i like this added validation, but i was thinking more along the lines of https://pkg.go.dev/github.com/docker/docker/client#Client.DistributionInspect . i could have explained it better. thanks for adding this.

Reply:
oh well instead of using DistributionInspect, i imported distribution and inspected it! lol 

to be fair i yoinked that entire helper function from go-vela/worker, and i kind of like that its tied to the general container spec and not docker

Reply:
true. i was thinking of actually pulling info from the registry (which is what [DistributionInspect](https://pkg.go.dev/github.com/docker/docker/client#Client.DistributionInspect) does vs a image ref format validation. we can add it later if we think that would be useful.